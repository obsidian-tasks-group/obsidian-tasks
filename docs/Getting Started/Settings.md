---
publish: true
---

# Settings

## Restart after changing settings

Regrettably, many of the settings for customising Tasks require one of the following to be done, for the new options to take effect:

- Obsidian to be restarted,
- or the vault to be closed and re-opened.

The Tasks Settings pane warns of this:

![Changing any settings requires a restart of obsidian](../images/settings-restart-after-change.png)

## Available settings

As the number of options in Tasks has grown, we have documented each setting in the page for the relevant topic.

For convenience, here is a list of all those documentation pages (in the order the options appear in the Tasks Settings pane):

- [[About Task Formats#Selecting the task format]]
- [[Global Filter]]
- [[Global Query]]
- [[Status Settings]]
- [[Dates#Date-tracking settings]]
- [[Recurring Tasks#Order of the new task]]
- [[Auto-Suggest]]
- [[Create or edit Task#Turning off keyboard shortcuts]]
