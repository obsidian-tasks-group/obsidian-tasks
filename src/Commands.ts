import { Editor, MarkdownView, Plugin, View, Workspace } from 'obsidian';

import { Task } from './Task';

export class Commands {
    private readonly plugin: Plugin;

    private get workspace(): Workspace {
        return this.plugin.app.workspace;
    }

    constructor({ plugin }: { plugin: Plugin }) {
        this.plugin = plugin;

        plugin.addCommand({
            id: 'toggle-done',
            name: 'Toggle Done',
            checkCallback: (checking: boolean) => {
                if (checking) {
                    const markdownView = this.getActiveMarkdownView();
                    if (markdownView === null) {
                        return false;
                    }

                    const editor = this.getEditor();
                    if (editor === null) {
                        // If we are not in an editor, the command shouldn't be shown.
                        return false;
                    }

                    const currentLine = editor.getLine(editor.getCursor().line);

                    // We want to toggle any checklist item, task or not, as this command is
                    // supposed to replace the original checklist toggle command.
                    // So we do not need to check whether we can create a valid task from the line.
                    const isTasksLine = Task.taskRegex.test(currentLine);

                    return isTasksLine;
                }

                // We are certain we are in the editor on a tasks line due to the check above.
                const path = (this.workspace.activeLeaf.view as any)?.file
                    ?.path;
                if (path === undefined) {
                    return;
                }
                // We are certain we are in the editor on a tasks line due to the check above.
                const editor = this.getEditor();
                if (editor === null) {
                    return;
                }

                const cursorPosition = editor.getCursor();
                const lineNumber = cursorPosition.line;
                const line = editor.getLine(lineNumber);
                const task = Task.fromLine({
                    line,
                    path,
                    sectionStart: 0, // We don't need this to toggle it here in the editor.
                    sectionIndex: 0, // We don't need this to toggle it here in the editor.
                    precedingHeader: null, // We don't need this to toggle it here in the editor.
                });

                if (task === null) {
                    // If the task is null this means that we have a regular checklist item.
                    // As this command is supposed to replace the original command to toggle
                    // checklists, we must do a regular toggle here.
                    const regexMatch = line.match(Task.taskRegex);
                    if (regexMatch === null) {
                        // We cannot toggle if we do not match.
                        return;
                    }

                    const indentation = regexMatch[1];
                    const statusString = regexMatch[2].toLowerCase();
                    const description = regexMatch[3];
                    const rest = regexMatch[4];

                    const toggledStatusString =
                        statusString === ' ' ? 'x' : ' ';

                    const toggledLine = `${indentation}- [${toggledStatusString}] ${description}${rest}`;
                    editor.setLine(lineNumber, toggledLine);
                } else {
                    // Toggle a regular task.
                    const toggledTask = task.toggle();
                    editor.setLine(lineNumber, toggledTask.toFileLineString());
                }
            },
        });
    }

    private getActiveMarkdownView(): MarkdownView | null {
        if (this.workspace.activeLeaf === undefined) {
            return null;
        }

        const activeLeaf = this.workspace.activeLeaf;
        if (!(activeLeaf.view instanceof MarkdownView)) {
            return null;
        }

        return activeLeaf.view;
    }

    private getEditor(): Editor | null {
        const view: View = this.workspace.activeLeaf.view;
        if (view instanceof MarkdownView) {
            return view.editor;
        } else {
            return null;
        }
    }
}
