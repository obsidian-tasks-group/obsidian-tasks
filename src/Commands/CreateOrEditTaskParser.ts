import { Status } from '../Status';
import { Priority, Task, TaskRegularExpressions } from '../Task';
import { DateFallback } from '../DateFallback';
import { StatusRegistry } from '../StatusRegistry';
import { TaskLocation } from '../TaskLocation';
import { getSettings } from '../Config/Settings';

/**
 * Read any markdown line and treat it as a task, for the purposes of
 * the 'Create or edit task' modal.
 *
 * Unlike {@link Task.fromLine}, which only processes tasks
 * already recognised by the Tasks plugin, this function processes any line.
 *
 * This is an implementation detail of that command, which has been separated
 * out to a different source file in order to allow its logic to be tested.
 *
 * @param line - The line the user had clicked on when running 'Create or edit task' command
 * @param path - The path of the file containing the line
 */
export const taskFromLine = ({ line, path }: { line: string; path: string }): Task => {
    const fallbackDate = DateFallback.fromPath(path);

    const task = Task.fromLine({
        line,
        taskLocation: TaskLocation.fromUnknownPosition(path), // We don't need precise location to toggle it here in the editor.
        fallbackDate, // set the scheduled date from the filename, so it can be displayed in the dialog
    });

    if (task !== null) {
        return task;
    }

    const { setCreatedDate } = getSettings();
    let createdDate: moment.Moment | null = null;
    if (setCreatedDate) {
        createdDate = window.moment();
    }

    // If we are not on a line of a task, we take what we have.
    // The non-task line can still be a checklist, for example if it is lacking the global filter.
    const nonTaskMatch = line.match(TaskRegularExpressions.nonTaskRegex);
    if (nonTaskMatch === null) {
        // Should never happen; everything in the regex is optional.
        console.error('Tasks: Cannot create task on line:', line);

        return new Task({
            status: Status.TODO,
            description: '',
            // We don't need the location fields except file to edit here in the editor.
            taskLocation: TaskLocation.fromUnknownPosition(path),
            indentation: '',
            listMarker: '-',
            priority: Priority.None,
            createdDate,
            startDate: null,
            scheduledDate: null,
            dueDate: null,
            doneDate: null,
            recurrence: null,
            blockLink: '',
            tags: [],
            reminder: null,
            originalMarkdown: '',
            scheduledDateIsInferred: false,
        });
    }

    const indentation: string = nonTaskMatch[1];
    const listMarker = nonTaskMatch[2] ?? '-';
    const statusString: string = nonTaskMatch[4] ?? ' ';
    const status = StatusRegistry.getInstance().bySymbolOrCreate(statusString);

    let description: string = nonTaskMatch[5];

    const blockLinkMatch = line.match(TaskRegularExpressions.blockLinkRegex);
    const blockLink = blockLinkMatch !== null ? blockLinkMatch[0] : '';

    if (blockLink !== '') {
        description = description.replace(TaskRegularExpressions.blockLinkRegex, '');
    }

    return new Task({
        status,
        description,
        // We don't need the location fields except file to edit here in the editor.
        taskLocation: TaskLocation.fromUnknownPosition(path),
        indentation,
        listMarker,
        blockLink,
        priority: Priority.None,
        createdDate,
        startDate: null,
        scheduledDate: null,
        dueDate: null,
        doneDate: null,
        recurrence: null,
        tags: [],
        reminder: null,
        originalMarkdown: '',
        // Not needed since the inferred status is always re-computed after submitting.
        scheduledDateIsInferred: false,
    });
};
