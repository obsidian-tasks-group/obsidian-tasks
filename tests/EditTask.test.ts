/**
 * @jest-environment jsdom
 */
import { fireEvent, render } from '@testing-library/svelte';
import { describe, expect, it } from '@jest/globals';
import moment from 'moment/moment';
import { taskFromLine } from '../src/Commands/CreateOrEditTaskParser';
import type { Task } from '../src/Task';
import EditTask from '../src/ui/EditTask.svelte';
import { Status } from '../src/Status';
import { DateFallback } from '../src/DateFallback';

window.moment = moment;
const task = taskFromLine({ line: '- [X] test task 1', path: '' });
const statusOptions: Status[] = [Status.DONE, Status.TODO];

describe('Task editing (UI)', () => {
    let resolvePromise: (input: string) => void;
    let waitForClose: Promise<string>;
    let onSubmit: (updatedTasks: Task[]) => void;

    beforeEach(() => {
        waitForClose = new Promise<string>((resolve, _) => {
            resolvePromise = resolve;
        });
        onSubmit = (updatedTasks: Task[]): void => {
            const serializedTask = DateFallback.removeInferredStatusIfNeeded(task, updatedTasks)
                .map((task: Task) => task.toFileLineString())
                .join('\n');
            resolvePromise(serializedTask);
        };
    });

    it('task description should be displayed', () => {
        const { container } = render(EditTask, { task, statusOptions, onSubmit });
        expect(() => container).toBeTruthy();
        const description = container.ownerDocument.getElementById('description') as HTMLInputElement;
        expect(() => description).toBeTruthy();
        expect(description!.value).toEqual('test task 1');
    });

    it('task description should be updated', async () => {
        const result = render(EditTask, { task, statusOptions, onSubmit });
        const { container } = result;
        expect(() => container).toBeTruthy();

        const description = container.ownerDocument.getElementById('description') as HTMLInputElement;
        expect(description).toBeTruthy();
        const submit = result.getByText('Apply') as HTMLButtonElement;
        expect(submit).toBeTruthy();

        await fireEvent.input(description, { target: { value: 'task edited' } });
        submit.click();
        const editedTask = await waitForClose;
        expect(editedTask).toEqual('- [X] task edited');
    });
});
