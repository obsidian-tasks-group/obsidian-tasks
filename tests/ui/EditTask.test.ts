/**
 * @jest-environment jsdom
 */
import { describe, expect, it } from '@jest/globals';
import { type RenderResult, fireEvent, render } from '@testing-library/svelte';
import moment from 'moment';
import { taskFromLine } from '../../src/Commands/CreateOrEditTaskParser';
import { GlobalFilter } from '../../src/Config/GlobalFilter';
import { resetSettings, updateSettings } from '../../src/Config/Settings';
import { StatusRegistry } from '../../src/Statuses/StatusRegistry';
import { DateFallback } from '../../src/Task/DateFallback';
import type { Task } from '../../src/Task/Task';
import EditTask from '../../src/ui/EditTask.svelte';
import { verifyWithFileExtension } from '../TestingTools/ApprovalTestHelpers';
import { verifyAllCombinations3Async } from '../TestingTools/CombinationApprovalsAsync';
import { prettifyHTML } from '../TestingTools/HTMLHelpers';
import { TaskBuilder } from '../TestingTools/TaskBuilder';

window.moment = moment;
/**
 * Construct an onSubmit function for editing the given task, and when Apply is clicked,
 * returning the edit task(s) converted to a string.
 * @param task
 */
function constructSerialisingOnSubmit(task: Task) {
    let resolvePromise: (input: string) => void;
    const waitForClose = new Promise<string>((resolve, _) => {
        resolvePromise = resolve;
    });

    const onSubmit = (updatedTasks: Task[]): void => {
        const serializedTask = DateFallback.removeInferredStatusIfNeeded(task, updatedTasks)
            .map((task: Task) => task.toFileLineString())
            .join('\n');
        resolvePromise(serializedTask);
    };

    return { waitForClose, onSubmit };
}

function renderAndCheckModal(task: Task, onSubmit: (updatedTasks: Task[]) => void, allTasks = [task]) {
    const result: RenderResult<EditTask> = render(EditTask, {
        task,
        statusOptions: StatusRegistry.getInstance().registeredStatuses,
        onSubmit,
        allTasks,
    });
    const { container } = result;
    expect(() => container).toBeTruthy();
    return { result, container };
}

/**
 * Find the element with the given id.
 * Template type T might be, for example, HTMLInputElement or HTMLSelectElement
 * @param container
 * @param elementId
 */
function getAndCheckRenderedElement<T>(container: HTMLElement, elementId: string) {
    const element = container.ownerDocument.getElementById(elementId) as T;
    expect(() => element).toBeTruthy();
    return element;
}

function getAndCheckRenderedDescriptionElement(container: HTMLElement): HTMLInputElement {
    return getAndCheckRenderedElement<HTMLInputElement>(container, 'description');
}

function getAndCheckApplyButton(result: RenderResult<EditTask>): HTMLButtonElement {
    const submit = result.getByText('Apply') as HTMLButtonElement;
    expect(submit).toBeTruthy();
    return submit;
}

async function editInputElement(inputElement: HTMLInputElement, newValue: string) {
    await fireEvent.input(inputElement, { target: { value: newValue } });
}

async function editInputElementAndSubmit(
    inputElement: HTMLInputElement,
    newValue: string,
    submit: HTMLButtonElement,
    waitForClose: Promise<string>,
): Promise<string> {
    await editInputElement(inputElement, newValue);
    submit.click();
    return await waitForClose;
}

function convertDescriptionToTaskLine(taskDescription: string): string {
    return `- [ ] ${taskDescription}`;
}

/**
 * Simulate the behaviour of:
 *   - clicking on a line in Obsidian,
 *   - opening the Edit task modal,
 *   - optionally editing the description,
 *   - and clicking Apply.
 * @param line
 * @param newDescription - the new value for the description field.
 *                         If `undefined`, the description won't be edited, unless text is needed to enable the Apply button.
 * @returns The edited task line.
 *
 * See also {@link editFieldAndSave} which is simpler, and works for more fields
 */
async function editTaskLine(line: string, newDescription: string | undefined) {
    const task = taskFromLine({ line: line, path: '' });
    const { waitForClose, onSubmit } = constructSerialisingOnSubmit(task);
    const { result, container } = renderAndCheckModal(task, onSubmit);

    const description = getAndCheckRenderedDescriptionElement(container);
    const submit = getAndCheckApplyButton(result);

    let adjustedNewDescription = newDescription ? newDescription : description!.value;
    if (!adjustedNewDescription) {
        adjustedNewDescription = 'simulate user typing text in to empty description field';
    }

    return await editInputElementAndSubmit(description, adjustedNewDescription, submit, waitForClose);
}

/**
 * Simulate the behaviour of:
 *   - clicking on a line in Obsidian,
 *   - opening the Edit task modal,
 *   - editing a field,
 *   - and clicking Apply.
 * @param line
 * @param elementId - specifying the field to edit
 * @param newValue - the new value for the field.
 * @returns The edited task line.
 *
 * See also {@link editTaskLine} which has extra logic for testing the description
 */
async function editFieldAndSave(line: string, elementId: string, newValue: string) {
    const task = taskFromLine({ line: line, path: '' });
    const { waitForClose, onSubmit } = constructSerialisingOnSubmit(task);
    const { result, container } = renderAndCheckModal(task, onSubmit);

    const description = getAndCheckRenderedElement<HTMLInputElement>(container, elementId);
    const submit = getAndCheckApplyButton(result);

    return await editInputElementAndSubmit(description, newValue, submit, waitForClose);
}

async function renderTaskModalAndChangeStatus(line: string, newStatusSymbol: string) {
    const task = taskFromLine({ line: line, path: '' });
    const { waitForClose, onSubmit } = constructSerialisingOnSubmit(task);
    const { result, container } = renderAndCheckModal(task, onSubmit);

    const statusSelector = getAndCheckRenderedElement<HTMLSelectElement>(container, 'status-type');
    const submit = getAndCheckApplyButton(result);

    await fireEvent.change(statusSelector, {
        target: { value: newStatusSymbol },
    });
    return { waitForClose, container, submit };
}

function getElementValue(container: HTMLElement, elementId: string) {
    const element = getAndCheckRenderedElement<HTMLInputElement>(container, elementId);
    return element.value;
}

describe('Task rendering', () => {
    afterEach(() => {
        GlobalFilter.getInstance().reset();
    });

    function testElementRender(line: string, elementId: string, expectedElementValue: string) {
        const task = taskFromLine({ line, path: '' });

        const onSubmit = (_: Task[]): void => {};
        const { container } = renderAndCheckModal(task, onSubmit);

        const inputElement = getAndCheckRenderedElement<HTMLInputElement>(container, elementId);
        expect(inputElement!.value).toEqual(expectedElementValue);
    }

    function testDescriptionRender(taskDescription: string, expectedDescription: string) {
        const line = convertDescriptionToTaskLine(taskDescription);
        testElementRender(line, 'description', expectedDescription);
    }

    it('should display task description (empty Global Filter)', () => {
        testDescriptionRender('important thing #todo', 'important thing #todo');
    });

    it('should display task description without non-tag Global Filter)', () => {
        GlobalFilter.getInstance().set('filter');
        testDescriptionRender('filter important thing', 'important thing');
    });

    it('should display task description with complex non-tag Global Filter)', () => {
        GlobalFilter.getInstance().set('filter');
        // This behavior is inconsistent with Obsidian's tag definition which includes nested tags
        testDescriptionRender('filter/important thing', 'filter/important thing');
    });

    it('should display task description without tag-like Global Filter', () => {
        GlobalFilter.getInstance().set('#todo');
        testDescriptionRender('#todo another plan', 'another plan');
    });

    it('should display task description with complex tag-like Global Filter', () => {
        GlobalFilter.getInstance().set('#todo');
        // This behavior is inconsistent with Obsidian's tag definition which includes nested tags
        testDescriptionRender('#todo/important another plan', '#todo/important another plan');
    });

    it('should display task description with emoji removed, when global filter is in initial line', () => {
        GlobalFilter.getInstance().set('#todo');
        testDescriptionRender(
            '#todo with global filter and with scheduled date ⏳ 2023-06-13',
            'with global filter and with scheduled date',
        );
    });

    it('should display task description with emoji removed, even if the global filter is missing from initial line (bug 2037)', () => {
        GlobalFilter.getInstance().set('#todo');
        // When written, this was a demonstration of the behaviour logged in
        // https://github.com/obsidian-tasks-group/obsidian-tasks/issues/2037
        testDescriptionRender(
            'without global filter but with scheduled date ⏳ 2023-06-13',
            'without global filter but with scheduled date', // fails, as the absence of the global filter means the line is not parsed, and the emoji stays in the description.
        );
    });

    const fullyPopulatedLine = TaskBuilder.createFullyPopulatedTask().toFileLineString();

    it('should display valid created date', () => {
        testElementRender(fullyPopulatedLine, 'created', '2023-07-01');
    });

    it('should display valid start date', () => {
        testElementRender(fullyPopulatedLine, 'start', '2023-07-02');
    });

    it('should display valid scheduled date', () => {
        testElementRender(fullyPopulatedLine, 'scheduled', '2023-07-03');
    });

    it('should display valid due date', () => {
        testElementRender(fullyPopulatedLine, 'due', '2023-07-04');
    });

    it('should display valid done date', () => {
        testElementRender(fullyPopulatedLine, 'done', '2023-07-05');
    });

    it('should display valid cancelled date', () => {
        testElementRender(fullyPopulatedLine, 'cancelled', '2023-07-06');
    });

    const invalidDateText = 'Invalid date';

    it('should display invalid cancelled date', () => {
        testElementRender('- [ ] ❌ 2024-02-31', 'cancelled', invalidDateText);
    });

    it('should display invalid created date', () => {
        testElementRender('- [ ] ➕ 2024-02-31', 'created', invalidDateText);
    });

    it('should display invalid done date', () => {
        testElementRender('- [ ] ✅ 2024-02-31', 'done', invalidDateText);
    });

    it('should display invalid due date', () => {
        testElementRender('- [ ] 📅 2024-02-31', 'due', invalidDateText);
    });

    it('should display invalid scheduled date', () => {
        testElementRender('- [ ] ⏳ 2024-02-31', 'scheduled', invalidDateText);
    });

    it('should display invalid start date', () => {
        testElementRender('- [ ] 🛫 2024-02-31', 'start', invalidDateText);
    });
});

describe('Task editing', () => {
    afterEach(() => {
        GlobalFilter.getInstance().reset();
    });

    async function testDescriptionEdit(taskDescription: string, newDescription: string, expectedDescription: string) {
        const line = convertDescriptionToTaskLine(taskDescription);
        const editedTask = await editTaskLine(line, newDescription);
        expect(editedTask).toEqual(`- [ ] ${expectedDescription}`);
    }

    it('should keep task description if it was not edited (Empty Global Filter)', async () => {
        const description = 'simple task #remember';
        await testDescriptionEdit(description, description, description);
    });

    it('should change task description if it was edited (Empty Global Filter)', async () => {
        await testDescriptionEdit('simple task #remember', 'another', 'another');
    });

    it('should not change the description if the task was not edited and keep Global Filter', async () => {
        const globalFilter = '#remember';
        const description = 'simple task';
        GlobalFilter.getInstance().set(globalFilter);
        await testDescriptionEdit(`${globalFilter} ${description}`, description, `${globalFilter} ${description}`);
    });

    it('should change the description if the task was edited and keep Global Filter', async () => {
        const globalFilter = '#remember';
        const oldDescription = 'simple task';
        const newDescription = 'new plan';
        GlobalFilter.getInstance().set(globalFilter);
        await testDescriptionEdit(
            `${globalFilter} ${oldDescription}`,
            newDescription,
            `${globalFilter} ${newDescription}`,
        );
    });

    describe('Status editing', () => {
        const today = '2024-02-29';
        beforeAll(() => {
            jest.useFakeTimers();
            jest.setSystemTime(new Date(today));
        });

        afterAll(() => {
            jest.useRealTimers();
        });

        afterEach(() => {
            resetSettings();
        });

        const line = '- [ ] simple';
        it('should change status to Done and add doneDate', async () => {
            const { waitForClose, container, submit } = await renderTaskModalAndChangeStatus(line, 'x');
            expect(getElementValue(container, 'done')).toEqual(today);

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot('"- [x] simple ✅ 2024-02-29"');
        });

        it('should change status to Todo and remove doneDate', async () => {
            const { waitForClose, container, submit } = await renderTaskModalAndChangeStatus(
                '- [x] simple ✅ 2024-02-29',
                ' ',
            );
            expect(getElementValue(container, 'done')).toEqual('');

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot('"- [ ] simple"');
        });

        it('should change status to Cancelled and add cancelledDate', async () => {
            const { waitForClose, container, submit } = await renderTaskModalAndChangeStatus(line, '-');
            expect(getElementValue(container, 'cancelled')).toEqual(today);

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot('"- [-] simple ❌ 2024-02-29"');
        });

        it('should create new instance of recurring task, with doneDate set to today', async () => {
            updateSettings({ recurrenceOnNextLine: false });
            const { waitForClose, submit } = await renderTaskModalAndChangeStatus(
                '- [ ] Recurring 🔁 every day 📅 2024-02-17',
                'x',
            );

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot(`
                "- [ ] Recurring 🔁 every day 📅 2024-02-18
                - [x] Recurring 🔁 every day 📅 2024-02-17 ✅ 2024-02-29"
            `);
        });

        it('should respect user setting for order of new recurring tasks', async () => {
            updateSettings({ recurrenceOnNextLine: true });
            const { waitForClose, submit } = await renderTaskModalAndChangeStatus(
                '- [ ] Recurring 🔁 every day 📅 2024-02-17',
                'x',
            );

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot(`
                "- [x] Recurring 🔁 every day 📅 2024-02-17 ✅ 2024-02-29
                - [ ] Recurring 🔁 every day 📅 2024-02-18"
            `);
        });

        it('should create new instance of "when done" recurring task, with doneDate set to today', async () => {
            updateSettings({ setCreatedDate: true });

            const { waitForClose, submit } = await renderTaskModalAndChangeStatus(
                '- [ ] Recurring 🔁 every day when done 📅 2024-02-17',
                'x',
            );

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot(`
                "- [ ] Recurring 🔁 every day when done ➕ 2024-02-29 📅 2024-03-01
                - [x] Recurring 🔁 every day when done 📅 2024-02-17 ✅ 2024-02-29"
            `);
        });

        it('should calculate the next recurrence date based on the actual done date in the field', async () => {
            updateSettings({ setCreatedDate: true });

            const { waitForClose, container, submit } = await renderTaskModalAndChangeStatus(
                '- [ ] Recurring 🔁 every day when done 📅 2024-02-17',
                'x',
            );

            const doneField = getAndCheckRenderedElement<HTMLInputElement>(container, 'done');
            await editInputElement(doneField, '2024-02-23');

            submit.click();
            expect(await waitForClose).toMatchInlineSnapshot(`
                "- [ ] Recurring 🔁 every day when done ➕ 2024-02-29 📅 2024-02-24
                - [x] Recurring 🔁 every day when done 📅 2024-02-17 ✅ 2024-02-23"
            `);
        });
    });

    describe('Date editing', () => {
        const line = '- [ ] simple';

        it('should edit and save cancelled date', async () => {
            expect(await editFieldAndSave(line, 'cancelled', '2024-01-01')).toEqual('- [ ] simple ❌ 2024-01-01');
        });

        it('should edit and save created date', async () => {
            expect(await editFieldAndSave(line, 'created', '2024-01-01')).toEqual('- [ ] simple ➕ 2024-01-01');
        });

        it('should edit and save done date', async () => {
            expect(await editFieldAndSave(line, 'done', '2024-01-01')).toEqual('- [ ] simple ✅ 2024-01-01');
        });

        it('should edit and save due date', async () => {
            expect(await editFieldAndSave(line, 'due', '2024-01-01')).toEqual('- [ ] simple 📅 2024-01-01');
        });

        it('should edit and save scheduled date', async () => {
            expect(await editFieldAndSave(line, 'scheduled', '2024-01-01')).toEqual('- [ ] simple ⏳ 2024-01-01');
        });

        it('should edit and save start date', async () => {
            expect(await editFieldAndSave(line, 'start', '2024-01-01')).toEqual('- [ ] simple 🛫 2024-01-01');
        });
    });
});

/**
 * @summary This tests behaviour under a wide variety of scenarios, such as multiple different user settings, and input lines.
 *
 * As the number of combinations of settings values increases, it becomes harder and harder
 * to write sufficient tests manually, and to find corner cases in exploratory testing.
 */
describe('Exhaustive editing', () => {
    beforeEach(() => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2023-07-18'));
    });

    afterEach(() => {
        GlobalFilter.getInstance().reset();
        resetSettings();
        jest.useRealTimers();
    });

    /**
     * Test outcome of simply editing and saving a task line, under many conditions.
     * Written as our previous test coverage was not good enough to detect the following:
     *   - https://github.com/obsidian-tasks-group/obsidian-tasks/issues/2112
     *   - Since Tasks 4.0.1, using 'Create or edit task' on a line with a checkbox
     *     but no global filter no longer adds the Created date.
     */
    describe('Edit and save', () => {
        const name = 'All inputs';
        const title = 'KEY: (globalFilter, set created date)\n';
        const globalFilterValues = ['', '#task'];
        const setCreatedDateValues = [false, true];
        const initialTaskLineValues = [
            '',
            'plain text, not a list item',
            '-',
            '- ',
            '- [ ]',
            '- [ ] ',
            '- list item, but no checkbox',
            '- [ ] checkbox with initial description',
            '- [ ] checkbox with initial description and created date ➕ 2023-01-01',
            '- [ ] #task checkbox with global filter string and initial description',
            '- [ ] checkbox with initial description ending with task tag at end #task',
        ];

        // For explanation of this call, see:
        // https://publish.obsidian.md/tasks-contributing/Testing/Approval+Tests#Verify+the+results+of+multiple+input+values
        verifyAllCombinations3Async<string, boolean, string>(
            name,
            title,
            async (globalFilter, setCreatedDate, initialTaskLine) => {
                GlobalFilter.getInstance().set(globalFilter as string);

                // @ts-expect-error: TS2322: Type 'T2' is not assignable to type 'boolean | undefined'.
                updateSettings({ setCreatedDate });

                // @ts-expect-error: TS2345: Argument of type 'T3' is not assignable to parameter of type 'string'.
                const editedTaskLine = await editTaskLine(initialTaskLine, undefined);

                return `
('${globalFilter}', ${setCreatedDate})
    '${initialTaskLine}' =>
    '${editedTaskLine}'`;
            },
            globalFilterValues,
            setCreatedDateValues,
            initialTaskLineValues,
        );
    });
});

describe('Edit Modal HTML snapshot tests', () => {
    afterEach(() => {
        resetSettings();
    });

    function verifyModalHTML() {
        const task = taskFromLine({ line: '- [ ] absolutely to do', path: '' });
        const onSubmit = () => {};
        const allTasks = [task];
        const { container } = renderAndCheckModal(task, onSubmit, allTasks);

        const prettyHTML = prettifyHTML(container.innerHTML);
        verifyWithFileExtension(prettyHTML, 'html');
    }

    it('should match snapshot', () => {
        updateSettings({ provideAccessKeys: true });
        verifyModalHTML();
    });

    it('should match snapshot - without access keys', () => {
        updateSettings({ provideAccessKeys: false });
        verifyModalHTML();
    });
});
