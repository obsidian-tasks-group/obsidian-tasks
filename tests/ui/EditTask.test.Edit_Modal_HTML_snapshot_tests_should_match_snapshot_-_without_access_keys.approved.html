<div>
  <form class="tasks-modal">
    <section class="tasks-modal-description-section">
      <label for="description">Description</label>
      <textarea id="description" class="tasks-modal-description" placeholder="Take out the trash"></textarea>
    </section>
    <section class="tasks-modal-priority-section">
      <label for="priority-none">Priority</label>
      <div class="task-modal-priority-option-container">
        <input type="radio" id="priority-lowest" value="lowest" />
        <label for="priority-lowest">
          <span>Lowest</span>
          <span>⏬</span>
        </label>
      </div>
      <div class="task-modal-priority-option-container">
        <input type="radio" id="priority-low" value="low" />
        <label for="priority-low">
          <span>Low</span>
          <span>🔽</span>
        </label>
      </div>
      <div class="task-modal-priority-option-container">
        <input type="radio" id="priority-none" value="none" />
        <label for="priority-none"><span>Normal</span></label>
      </div>
      <div class="task-modal-priority-option-container">
        <input type="radio" id="priority-medium" value="medium" />
        <label for="priority-medium">
          <span>Medium</span>
          <span>🔼</span>
        </label>
      </div>
      <div class="task-modal-priority-option-container">
        <input type="radio" id="priority-high" value="high" />
        <label for="priority-high">
          <span>High</span>
          <span>⏫</span>
        </label>
      </div>
      <div class="task-modal-priority-option-container">
        <input type="radio" id="priority-highest" value="highest" />
        <label for="priority-highest">
          <span>Highest</span>
          <span>🔺</span>
        </label>
      </div>
    </section>
    <hr />
    <section class="tasks-modal-dates-section">
      <label for="recurrence">Recurs</label>
      <input id="recurrence" type="text" class="tasks-modal-date-input" placeholder="Try 'every day when done'" />
      <code class="tasks-modal-parsed-date">
        🔁
        <i>not recurring</i>
      </code>
      <label for="due">Due</label>
      <input id="due" type="text" class="tasks-modal-date-input" placeholder="Try 'Mon' or 'tm' then space" />
      <code class="tasks-modal-parsed-date">
        📅
        <i>no due date</i>
      </code>
      <label for="scheduled">Scheduled</label>
      <input id="scheduled" type="text" class="tasks-modal-date-input" placeholder="Try 'Mon' or 'tm' then space" />
      <code class="tasks-modal-parsed-date">
        ⏳
        <i>no scheduled date</i>
      </code>
      <label for="start">Start</label>
      <input id="start" type="text" class="tasks-modal-date-input" placeholder="Try 'Mon' or 'tm' then space" />
      <code class="tasks-modal-parsed-date">
        🛫
        <i>no start date</i>
      </code>
      <div class="future-dates-only">
        <label for="forwardOnly">Only future dates:</label>
        <input id="forwardOnly" type="checkbox" class="task-list-item-checkbox tasks-modal-checkbox" />
      </div>
    </section>
    <hr />
    <section class="tasks-modal-dependencies-section">
      <label for="blockedBy">Before this</label>
      <span>
        <input
          id="blockedBy"
          class="tasks-modal-dependency-input"
          type="text"
          placeholder="Search for tasks that the task being edited depends on..." />
        <iframe
          style="
            display: block;
            position: absolute;
            top: 0;
            left: 0;
            width: 100%;
            height: 100%;
            overflow: hidden;
            border: 0;
            opacity: 0;
            pointer-events: none;
            z-index: -1;
          "
          aria-hidden="true"
          tabindex="-1"
          src="about:blank"></iframe>
      </span>
      <label for="blocking">After this</label>
      <span>
        <input
          id="blocking"
          class="tasks-modal-dependency-input"
          type="text"
          placeholder="Search for tasks that depend on this task being done..." />
        <iframe
          style="
            display: block;
            position: absolute;
            top: 0;
            left: 0;
            width: 100%;
            height: 100%;
            overflow: hidden;
            border: 0;
            opacity: 0;
            pointer-events: none;
            z-index: -1;
          "
          aria-hidden="true"
          tabindex="-1"
          src="about:blank"></iframe>
      </span>
    </section>
    <hr />
    <section class="tasks-modal-dates-section">
      <label for="status">Status</label>
      <select id="status-type" class="status-editor-status-selector">
        <option value=" ">Todo [ ]</option>
        <option value="/">In Progress [/]</option>
        <option value="x">Done [x]</option>
        <option value="-">Cancelled [-]</option>
      </select>
      <label for="created">Created</label>
      <input id="created" type="text" class="tasks-modal-date-input" placeholder="Try 'Mon' or 'tm' then space" />
      <code class="tasks-modal-parsed-date">
        ➕
        <i>no created date</i>
      </code>
      <label for="done">Done</label>
      <input id="done" type="text" class="tasks-modal-date-input" placeholder="Try 'Mon' or 'tm' then space" />
      <code class="tasks-modal-parsed-date">
        ✅
        <i>no done date</i>
      </code>
      <label for="cancelled">Cancelled</label>
      <input id="cancelled" type="text" class="tasks-modal-date-input" placeholder="Try 'Mon' or 'tm' then space" />
      <code class="tasks-modal-parsed-date">
        ❌
        <i>no cancelled date</i>
      </code>
    </section>
    <section class="tasks-modal-button-section">
      <button type="submit" class="mod-cta">Apply</button>
      <button type="button">Cancel</button>
    </section>
  </form>
</div>
